ethcore/src/spec/spec.rs

```diff
@@ -119,39 +121,42 @@ pub struct Spec {
 	constructors: Vec<(Address, Bytes)>,
 
 	/// May be prepopulated if we know this in advance.
-	state_root_memo: RwLock<Option<H256>>,
+	state_root_memo: RwLock<H256>,
 
 	/// Genesis state as plain old data.
 	genesis_state: PodState,
 }
 
-impl From<ethjson::spec::Spec> for Spec {
-	fn from(s: ethjson::spec::Spec) -> Self {
-		let builtins = s.accounts.builtins().into_iter().map(|p| (p.0.into(), From::from(p.1))).collect();
-		let g = Genesis::from(s.genesis);
-		let GenericSeal(seal_rlp) = g.seal.into();
-		let params = CommonParams::from(s.params);
-		Spec {
-			name: s.name.clone().into(),
-			params: params.clone(),
-			engine: Spec::engine(s.engine, params, builtins),
-			data_dir: s.data_dir.unwrap_or(s.name).into(),
-			nodes: s.nodes.unwrap_or_else(Vec::new),
-			parent_hash: g.parent_hash,
-			transactions_root: g.transactions_root,
-			receipts_root: g.receipts_root,
-			author: g.author,
-			difficulty: g.difficulty,
-			gas_limit: g.gas_limit,
-			gas_used: g.gas_used,
-			timestamp: g.timestamp,
-			extra_data: g.extra_data,
-			seal_rlp: seal_rlp,
-			constructors: s.accounts.constructors().into_iter().map(|(a, c)| (a.into(), c.into())).collect(),
-			state_root_memo: RwLock::new(g.state_root),
-			genesis_state: From::from(s.accounts),
-		}
-	}
+fn load_from(s: ethjson::spec::Spec) -> Result<Spec, Error> {
+	let builtins = s.accounts.builtins().into_iter().map(|p| (p.0.into(), From::from(p.1))).collect();
+	let g = Genesis::from(s.genesis);
+	let GenericSeal(seal_rlp) = g.seal.into();
+	let params = CommonParams::from(s.params);
+
+	let s = Spec {
+		name: s.name.clone().into(),
+		params: params.clone(),
+		engine: Spec::engine(s.engine, params, builtins),
+		data_dir: s.data_dir.unwrap_or(s.name).into(),
+		nodes: s.nodes.unwrap_or_else(Vec::new),
+		parent_hash: g.parent_hash,
+		transactions_root: g.transactions_root,
+		receipts_root: g.receipts_root,
+		author: g.author,
+		difficulty: g.difficulty,
+		gas_limit: g.gas_limit,
+		gas_used: g.gas_used,
+		timestamp: g.timestamp,
+		extra_data: g.extra_data,
+		seal_rlp: seal_rlp,
+		constructors: s.accounts.constructors().into_iter().map(|(a, c)| (a.into(), c.into())).collect(),
+		state_root_memo: RwLock::new(Default::default()), // will be overwritten right after.
+		genesis_state: s.accounts.into(),
+	};
+
+	let _ = s.run_constructors(&Default::default(), BasicBackend(MemoryDB::new()))?;
```

Comment:
I'd rather put an optional pre-computed state root into the json. Alternatively light client could cache the genesis header. Full client only needs to call `spec.genesis_header()` when the database is empty IIRC.